Hook up the tests via cmake and in the github workflows

whot opened the pull request:
github workflow is for the linux builds only atm. And obviously they fail because of #1490, so waiting with this PR until #1494 is merged is best.

p12tic commented on clean_build.sh:
```diff
@@ -4,26 +4,23 @@ cd "$(dirname "$0")" || exit 1
-else
-    CMAKE_HAS_PARALLEL=0
+# Let's be reasonably nice in shared envs like CI runners
+nproc=$(nproc)
```
This script is also used on OSX, `nproc` is not available there.

whot replied:
fixed by running `sysctl` on darwin

p12tic commented on clean_build.sh:
```diff
@@ -4,26 +4,23 @@ cd "$(dirname "$0")" || exit 1
-        exit 1
-    fi
+[ -n "$B_CMAKE" ] || B_CMAKE=$(command -v cmake3)
+[ -n "$B_CMAKE" ] || B_CMAKE=$(command -v cmake2)
```
`cmake2` will probably refer to cmake v2.x if present. This won't work, because the project has `cmake_minimum_required (VERSION 3.4). I agree with the change of checking style otherwise.

whot replied:
Fixed!

p12tic commented on clean_build.sh:
```diff
@@ -41,23 +38,15 @@ fi
-echo "Starting Input Leap $B_BUILD_TYPE build..."
-"$B_CMAKE" $B_CMAKE_FLAGS .. || exit 1
-
-if [ "$CMAKE_HAS_PARALLEL" -eq "1" ]; then
```
This complication here is to support building on older distributions. E.g. Ubuntu 18.04 only has cmake 3.10, whereas `--parallel` and friends became supported since 3.12.

whot replied:
yep, but [the documentation](https://cmake.org/cmake/help/latest/manual/cmake.1.html#build-tool-mode) points at `CMAKE_BUILD_PARALLEL_LEVEL` being the override when `--parallel` is not given:

> The [CMAKE_BUILD_PARALLEL_LEVEL](https://cmake.org/cmake/help/latest/envvar/CMAKE_BUILD_PARALLEL_LEVEL.html#envvar:CMAKE_BUILD_PARALLEL_LEVEL) environment variable, if set, specifies a default parallel level when this option is not given.

iow, we no longer need the check for versions because old cmake won't care about some random env var and new cmake handles it correctly.

p12tic replied:
Indeed, your solution is equivalent to the previous code in less lines. I missed that `ninja` would run using all cores anyway.

whot replied:
> I missed that ninja would run using all cores anyway.

so, interestingly enough: we're now actually *limiting* the number of cores to be used. I picked 32 as arbitrary max to be nice on beefy runners (on gitlab.fdo we have `FDO_CI_CONCURRENT` that specifies how many you can use in a job).

If we just want to run with whatever is the default, we can drop this code altogether because then ninja will just use whatever its built-in default is. We could export `MAKEFLAGS=-j` to ensure make does run in parallel where we fall back to that. but for ninja, no work is needed.

p12tic replied:
That's good point, we don't need unnecessary limiting. Myself I thought the reason you wanted that is to have a limit on some CI. Given that 32 threads is plenty for such a small project, I didn't think much about this limitation. But if there's no need for it after all, let's remove it.

whot replied:
Done in #1502

whot commented:
Fixed the comments above (and the commit message which still referred to changes I had to drop since). I dropped the commit to add the test suite to the github workflow though because right now it fails and let's not have a broken pipeline until we can figure out what the test case failure is. I filed #1497 for that commit now.
